File: Python/t19.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 给定一个链表，删除链表的倒数第 n 个节点，并且返回链表的头结点。
#
# 示例：
# 给定一个链表: 1->2->3->4->5, 和 n = 2.
#
# 当删除了倒数第二个节点后，链表变为 1->2->3->5.
# 说明：
# 给定的 n 保证是有效的。
#
# 进阶：
# 你能尝试使用一趟扫描实现吗？



# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# 解法一：双循环
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:

        # 创建一个新的头节点用于后续删除节点后返回
        new_head = ListNode(0)
        new_head.next = head

        # 计算链表的长度
        nodeLens = 0
        while head:
            nodeLens += 1
            head = head.next

        # 计算从头到删除节点的长度
        subLens = nodeLens - n

        # 创建新的头节点p用于删除对应节点
        p = new_head

        while subLens > 0:
            subLens -= 1
            p = p.next

        # 遍历到删除的节点后将next指针指向下下个元素的next
        p.next = p.next.next

        return new_head.next



    def showNode(self, node : ListNode) -> list:
        """
        show all value of ListNode
        :param node:
        :return:
        """
        nodeValue = []
        head = node
        while head:
            nodeValue.append(head.val)
            head = head.next
        return nodeValue




# 解法二：双指针
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:

        # 创建一个新的头节点用于后续删除节点后返回
        new_head = ListNode(0)
        new_head.next = head

        # 设置双指针
        first = new_head
        second = new_head

        # 第一个指针先走n步
        for i in range(n+1):
            first = first.next

        # 接着上一步，两个指针同时走，直到第一个指针到了尾部
        while first != None:
            first = first.next
            second = second.next
        # 第二个指针与第一个指针相差n步，这时跳过下一个元素即可
        second.next = second.next.next

        return new_head.next



    def showNode(self, node : ListNode) -> list:
        """
        show all value of ListNode
        :param node:
        :return:
        """
        nodeValue = []
        head = node
        while head:
            nodeValue.append(head.val)
            head = head.next
        return nodeValue




node = ListNode(1)
node.next = ListNode(2)
node.next.next = ListNode(3)
node.next.next.next = ListNode(4)
node.next.next.next.next = ListNode(5)

result = Solution()
n = 2

new_node = result.removeNthFromEnd(node, n)
print(result.showNode(new_node))

File: C++/t19.cpp
#include <cstddef>
#include <iostream>

using namespace std;

/**
 *  leetcode第19题
 *  利用双指针移动步长差求解
 *  Author:lyc
*/


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {

        // 创建一个保护节点，不用对边界进行特殊处理
        ListNode* newNode = new ListNode(0);

        // 保护节点与head连接
        newNode->next = head;

        // 创建两个指针用于移动
        ListNode* first = newNode;
        ListNode* second = newNode;

        // 先移动第一个指针，移动n个步长
        for(int i=0; i <= n ; i++)
        {
            first = first->next;
        }

        // 两个指针同时移动，直到第一个指针为NULL
        while (first != NULL)
        {
            first = first->next;
            second = second->next;
        }

        // 此时两个指针相差n步，第二个指针的下一个元素即为要删除的元素，跳过此元素
        second->next = second->next->next;

        // 返回链表时跳过保护节点
        return newNode->next;
        
    };
};


int main()
{
    /**
     * 利用定义好的链表结构体创建一个链表
    */
    ListNode* node1 = new ListNode(1);
    ListNode* node2 = new ListNode(2);
    ListNode* node3 = new ListNode(3);
    ListNode* node4 = new ListNode(4);
    ListNode* node5 = new ListNode(5);

    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    node4->next = node5;
    
    // 声明需要删除倒数第几个元素
    int n = 2;

    Solution result;

    ListNode *res = result.removeNthFromEnd(node1, n);

    // 循环打印删除对应元素之后的链表
    while(res != NULL)
    {
        cout << res->val << endl;
        res = res->next;
    }

    system("pause");
    return 0;

}
